feat: add What's new in Kotlin 2.1.0 (#4574)
JetBrains/kotlin-web-site: docs/topics/whatsnew21.md

* **New language features in preview**: [Guard conditions in `when` with a subject](#guard-conditions-in-when-with-a-subject),
  [non-local `break` and `continue`](#non-local-break-and-continue), and [multi-dollar string interpolation](#multi-dollar-string-interpolation).
* **K2 compiler updates**: [More flexibility around compiler checks](#extra-compiler-checks) and [improvements to the kapt implementation](#improved-k2-kapt-implementation).
* **Kotlin Multiplatform**: Introduced [basic support for Swift export](#basic-support-for-swift-export),
  [stable Gradle DSL for compiler options](#new-gradle-dsl-for-compiler-options-in-multiplatform-projects-promoted-to-stable), and more.
* **Kotlin/Native**: [Improved support for `iosArm64`](#iosarm64-promoted-to-tier-1) and other updates.
* **Kotlin/Wasm**: Multiple updates, including [support for incremental compilation](#support-for-incremental-compilation).
* **Gradle support**: [Improved compatibility with newer versions of Gradle and the Android Gradle plugin](#gradle-improvements),
  along with [updates to the Kotlin Gradle plugin API](#new-api-for-kotlin-gradle-plugin-extensions).

> * **Documentation**: [Significant improvements to the Kotlin documentation](#documentation-updates).

Author: It is still "Documentation: Kotlin documentation".
I personally see no problem here 🤔 .

See [Update to a new Kotlin version](releases.md#update-to-a-new-kotlin-version) for details.

## Language

After the Kotlin 2.0.0 release with the K2 compiler, the JetBrains team is focusing on improving the language with new features.
In this release, we are excited to announce several new language design improvements.

These features are available in preview, and we encourage you to try them and share your feedback:

* [Guard conditions in `when` with a subject](#guard-conditions-in-when-with-a-subject)
* [Non-local `break` and `continue`](#non-local-break-and-continue)
* [Multi-dollar interpolation: improved handling of `$` in string literals](#multi-dollar-string-interpolation)

> All the features have IDE support in the latest 2024.3 version of IntelliJ IDEA with K2 mode enabled.
>
> Learn more in the [IntelliJ IDEA 2024.3 blog post](https://blog.jetbrains.com/idea/2024/11/intellij-idea-2024-3/).
>
{type="tip"}

[See the full list of Kotlin language design features and proposals](kotlin-language-features-and-proposals.md).

This release also brings the following language updates:

> * [](#support-for-requiring-opt-in-to-extend-apis)

Author: No, if there is no content in the `[]` the title of the anchor/article will be used.

```bash
kotlinc -Xnon-local-break-continue main.kt
```

Or add it in the `compilerOptions {}` block of your Gradle build file:

```kotlin
// build.gradle.kts
kotlin {
    compilerOptions {
        freeCompilerArgs.add("-Xnon-local-break-continue")
    }
}
```

We're planning to make this feature Stable in future Kotlin releases. 
If you encounter any issues when using non-local `break` and `continue`, 
please report them to our [issue tracker](https://youtrack.jetbrains.com/issue/KT-1436).

### Multi-dollar string interpolation

> The feature is [In preview](kotlin-evolution-principles.md#pre-stable-features)
> and opt-in is required (see details below).
> 
> We would appreciate your feedback in [YouTrack](https://youtrack.jetbrains.com/issue/KT-2425).
>
{style="warning"}

Kotlin 2.1.0 introduces support for multi-dollar string interpolation, 
improving how the dollar sign (`$`) is handled within string literals.
This feature is helpful in contexts that require multiple dollar signs,
such as templating engines, JSON schemas, or other data formats.

String interpolation in Kotlin uses a single dollar sign. 

> However, using a literal dollar sign in a string, 

Author: For what reason should we introduce the dollar sign again?
It was introduced in the previous para.

and another has a lambda function that returns a generic type `V`.
Similarly, there are two overloads for the extension function: `storeExtension()`.

When the `store()` function was called with and without a lambda function,
the compiler successfully resolved the correct overloads.
However, when the extension function `storeExtension()` was called with a lambda function,
the compiler didn't resolve the correct overload because it incorrectly considered both overloads to be applicable.

To fix this problem, we've introduced a new heuristic so that the compiler can discard a possible overload
when a function parameter with a generic type can't accept a lambda function based on information from a different argument.
This change makes the behavior of member functions and extension functions consistent,
and it is enabled by default in Kotlin 2.1.0.

### Improved exhaustiveness checks for when expressions with sealed classes

In previous versions of Kotlin, the compiler required an `else` branch in `when`
expressions for type parameters with sealed upper bounds, even when all cases in the `sealed class` hierarchy were covered.

> This behavior is addressed and improved in Kotlin 2.1.0,

Reviewer: ```suggestion

This behavior is addressed and improved in Kotlin 2.1.0,
```

Author: Well, not sure about this change.
Could you provide the motivation behind it? All other cases like this one remain the same...

| `REDUNDANT_CALL_OF_CONVERSION_METHOD`                 | `42.toInt()` is used instead of `42`                                                     |
| `USELESS_CALL_ON_NOT_NULL`                            | `"".orEmpty()` is used instead of `""`                                                   |
| `REDUNDANT_SINGLE_EXPRESSION_STRING_TEMPLATE`         | `"$string"` is used instead of `string`                                                  |
| `UNUSED_ANONYMOUS_PARAMETER`                          | A parameter is passed in the lambda expression but never used                            |
| `REDUNDANT_VISIBILITY_MODIFIER`                       | `public class Klass` is used instead of `class Klass`                                    |
| `REDUNDANT_MODALITY_MODIFIER`                         | `final class Klass` is used instead of `class Klass`                                     |
| `REDUNDANT_SETTER_PARAMETER_TYPE`                     | `set(value: Int)` is used instead of `set(value)`                                        |
| `CAN_BE_VAL`                                          | `var local = 0` is defined but never reassigned, can be `val local = 42` instead         |

> | `ASSIGNED_VALUE_IS_NEVER_READ`                        | `val local = 42` is defined but never used afterward in the code                         |

Reviewer: ```suggestion
| `ASSIGNED_VALUE_IS_NEVER_READ`                        | `val local = 42` is defined but never used in the code                         |
```

Reviewer: For conciseness and consitency with the next one

Reviewer 2: No, there's a difference. Otherwise, they are identical and there would be no need for two different warnings

Author: Not sure about this one, otherwise, the idea of the check is not clear.

| `UNUSED_VARIABLE`                                     | `val local = 0` is defined but never used in the code                                    |
| `REDUNDANT_RETURN_UNIT_TYPE`                          | `fun foo(): Unit {}` is used instead of `fun foo() {}`                                   |

> | `UNREACHABLE_CODE`                                    | Code statement is present but can never be executed                                      |

Reviewer: ```suggestion
| `UNREACHABLE_CODE`                                    | The code statement is present but can never be executed                                      |
```

Author: Skip it to provide more space in the table.

syntax in the command line or the `freeCompilerArgs` attribute in the `compilerOptions {}` block of your build file.

For example, if you have [extra compiler checks](#extra-compiler-checks) enabled in your project but want to suppress one of them, use:

```kotlin
// build.gradle.kts
kotlin {
    compilerOptions {
        extraWarnings.set(true)
        freeCompilerArgs.add("-Xsuppress-warning=CAN_BE_VAL")
    }
}
```

If you want to suppress a warning but don't know its name, select the element and click the light bulb icon (or use <shortcut>Cmd + Enter</shortcut>/<shortcut>Alt + Enter</shortcut>):

![Warning name intention](warning-name-intention.png){width=500}

The new compiler option is currently [Experimental](components-stability.md#stability-levels-explained).
The following details are also worth noting:

* Error suppression is not allowed.
* If you pass an unknown warning name, compilation will result in an error.
* You can specify several warnings at once:
  
   <tabs>
   <tab title="Command line">

   ```bash
   kotlinc -Xsuppress-warning=NOTHING_TO_INLINE -Xsuppress-warning=NO_TAIL_CALLS_FOUND main.kt
   ```

   </tab>
   <tab title="Build file">

   ```kotlin

>     // build.gradle.kts

Reviewer: ```suggestion
   // build.gradle.kts
```

* `strict`: Report errors for nullability mismatches (default mode).

For more information, see [Nullability annotations](java-interop.md#nullability-annotations).

## Kotlin Multiplatform

Kotlin 2.1.0 introduces [basic support for Swift export](#basic-support-for-swift-export) and makes
[publishing Kotlin Multiplatform libraries easier](#ability-to-publish-kotlin-libraries-from-any-host).
It also focuses on improvements around Gradle that stabilize the [new DSL for configuring compiler options](#new-gradle-dsl-for-compiler-options-in-multiplatform-projects-promoted-to-stable)
and bring a [preview of the Isolated Projects feature](#preview-gradle-s-isolated-projects-in-kotlin-multiplatform).

### New Gradle DSL for compiler options in multiplatform projects promoted to Stable

In Kotlin 2.0.0, [we introduced a new Experimental Gradle DSL](whatsnew20.md#new-gradle-dsl-for-compiler-options-in-multiplatform-projects)
to simplify the configuration of compiler options across your multiplatform projects.
In Kotlin 2.1.0, this DSL has been promoted to Stable.

The overall project configuration now has three layers. The highest is the extension level, 

> then the target level and the lowest is the compilation unit (which is usually a compilation task):

Reviewer: ```suggestion
then the target level, and the lowest is the compilation unit (which is usually a compilation task):
```

> 
> We would appreciate your feedback on it in [YouTrack](https://youtrack.jetbrains.com/issue/KT-57279/Support-Gradle-Project-Isolation-Feature-for-Kotlin-Multiplatform). 
> Opt-in is required (see details below).
>
{style="warning"}

In Kotlin 2.1.0, 
you can preview Gradle's [Isolated Projects](https://docs.gradle.org/current/userguide/isolated_projects.html) 
feature in your multiplatform projects.

The Isolated Projects feature in Gradle improves build performance by "isolating"
configuration of individual Gradle projects from each other.
Each project's build logic is restricted from directly accessing the mutable state of other projects,
allowing them to safely run in parallel.
To support this feature, we made some changes to the Kotlin Gradle plugin's model,
and we are interested in hearing about your experiences during this preview phase.

There are two ways to enable the Kotlin Gradle plugin's new model:

Option 1: **Testing compatibility without enabling Isolated Projects** –
To check compatibility with the Kotlin Gradle plugin's new model without enabling the Isolated Projects feature,

> add the following Gradle property in the `build.gradle.kts` file of your project:

Reviewer: ```suggestion
add the following Gradle property in the `gradle.properties` file of your project:
```

which you can access from Xcode.

#### How to enable Swift export

This feature is currently [Experimental](components-stability.md#stability-levels-explained).
To try it out in your project, add the following Gradle option to your `gradle.properties` file:

```none
# gradle.properties
kotlin.experimental.swift-export.enabled=true
```

#### Leave feedback on Swift export

We're planning to expand and stabilize Swift export support in future Kotlin releases.
Please leave your feedback in [this YouTrack issue](https://youtrack.jetbrains.com/issue/KT-64572).

### Ability to publish Kotlin libraries from any host

> This feature is currently [Experimental](components-stability.md#stability-levels-explained).
> Opt-in is required (see the details below), and you should use it only for evaluation purposes.
> We would appreciate your feedback on it in [YouTrack](https://youtrack.jetbrains.com/issue/KT-71290).
>
{style="warning"}

The Kotlin compiler produces `.klib` artifacts for publishing Kotlin libraries.
Previously, you could get the necessary artifacts from any host, except for Apple platform targets that required a Mac machine. 

> That put a special restraint on Kotlin Multiplatform projects that targeted iOS, macOS, tvOS, and watchOS targets.

Reviewer: ```suggestion
That puts a special restraint on Kotlin Multiplatform projects that targeted iOS, macOS, tvOS, and watchOS targets.
```

Author: Since we used "previously", it means "in the past". No change needed.

output type to cache the results of the task.

This should resolve issues where `UP-TO-DATE`
checks failed to detect changes to header files specified in the [definition file](native-definition-file.md),
preventing the build system from recompiling the code.

### Deprecation of the mimalloc memory allocator

Back in Kotlin 1.9.0, we introduced the new memory allocator, and then we enabled it by default in Kotlin 1.9.20.
The new allocator has been designed to make garbage collection more efficient
and improve the Kotlin/Native memory manager's runtime performance.

The new memory allocator replaced the previous default allocator, [mimalloc](https://github.com/microsoft/mimalloc).
Now, it's time to deprecate mimalloc in the Kotlin/Native compiler.

You can now remove the `-Xallocator=mimalloc` compiler option from your build scripts.
If you encounter any issues, please report them to our [issue tracker](http://kotl.in/issue).

For more information on the memory allocator and garbage collection in Kotlin,
see [Kotlin/Native memory management](native-memory-manager.md).

## Kotlin/Wasm

> Kotlin/Wasm received some multiple updates along with [support for incremental compilation](#support-for-incremental-compilation).

Reviewer: ```suggestion
Kotlin/Wasm received multiple updates along with [support for incremental compilation](#support-for-incremental-compilation).
```

    }
}
```

### Improved debugging experience for Kotlin/Wasm

Previously, when debugging Kotlin/Wasm code in web browsers, you might have encountered
a low-level representation of variable values in the debugging interface.
This often made it challenging to track the current state of the application.

![Kotlin/Wasm old debugger](wasm-old-debugger.png){width=700}

To improve this experience, custom formatters have been added in the variable view.
The implementation uses the [custom formatters API](https://firefox-source-docs.mozilla.org/devtools-user/custom_formatters/index.html),
which is supported across major browsers like Firefox and Chromium-based ones.

With this change, you can now display and locate variable values in a more user-friendly and comprehensible manner.

![Kotlin/Wasm improved debugger](wasm-debugger-improved.png){width=700}

To try the new debugging experience:

1. Add the following compiler option to the `wasmJs {}` compiler options:

   ```kotlin
   // build.gradle.kts
   kotlin {
       wasmJs {
           // ...
   
           compilerOptions {
               freeCompilerArgs.add("-Xwasm-debugger-custom-formatters")
           }
       }
   }
   ```

> 2. Enable custom formatters in your browser.

Reviewer: ```suggestion
2. Enable custom formatters in your browser:
```

>    In Chrome DevTools, it's available via **Settings | Preferences | Console**:

Reviewer: ```suggestion
   * In Chrome DevTools, it's available via **Settings | Preferences | Console**:
```

   ![Enable custom formatters in Chrome](wasm-custom-formatters-chrome.png){width=700}

>    In Firefox DevTools, it's available via **Settings | Advanced settings**:

Reviewer: ```suggestion
   * In Firefox DevTools, it's available via **Settings | Advanced settings**:
```

   ![Enable custom formatters in Firefox](wasm-custom-formatters-firefox.png){width=700}

### Reduced size of Kotlin/Wasm binaries

The size of your Wasm binaries produced by production builds will be reduced by up to 30%,
and you may see some performance improvements.
This is because the `--closed-world`, `--type-ssa`, and `--type-merging`
Binaryen options are now considered safe to use for all Kotlin/Wasm projects and are enabled by default.

### Improved JavaScript array interoperability in Kotlin/Wasm

While Kotlin/Wasm's standard library provides the `JsArray<T>` type for JavaScript arrays,

> there was no direct method to transform `JsArray<T>` into Kotlin's native `Array` or `List` types.

Reviewer: ```suggestion
there was no direct method to convert `JsArray<T>` into Kotlin's native `Array` or `List` types.
```

This change marks the last phase of a deprecation cycle to migrate to named exports:

**In version 2.0.0:** A warning message was printed to the console, explaining that exporting entities via default exports is deprecated.

**In version 2.0.20:** An error occurred, requesting the use of the corresponding named import.

**In version 2.1.0:** The use of default imports has been completely removed.

### Subproject-specific Node.js settings

You can configure Node.js settings for your project by defining properties of the `NodeJsRootPlugin` class for `rootProject`.
In 2.1.0, you can configure these settings for each subproject using a new class, `NodeJsPlugin`.
Here's an example demonstrating how to set a specific Node.js version for a subproject:

```kotlin
// build.gradle.kts
project.plugins.withType<org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsPlugin> {
    project.the<org.jetbrains.kotlin.gradle.targets.js.nodejs.NodeJsEnvSpec>().version = "22.0.0"
}
```

To use the new class for the entire project, add the same code in the `allprojects {}` block:

> ```kotlin

Reviewer: Are we missing some indentation here?

Author: What kind of?
Seems valid for me.

```kotlin
object Bar {
    val `property example`: String = "bar"
}

@JsQualifier("fooNamespace")
external object Foo {
    val `property example`: String
}

@JsExport
object Baz {
    val `property example`: String = "bar"
}

fun main() {
    // In JavaScript, this is compiled into Bar.property_example_HASH
    println(Bar.`property example`)
    // In JavaScript, this is compiled into fooNamespace["property example"]
    println(Foo.`property example`)
    // In JavaScript, this is compiled into Baz["property example"]
    println(Baz.`property example`)
}
```

### Support for generating ES2015 arrow functions

In Kotlin 2.1.0, Kotlin/JS introduces support for generating ES2015 arrow functions,
such as `(a, b) => expression`, instead of anonymous functions.

Using arrow functions can reduce the bundle size of your project,
especially when using the experimental `-Xir-generate-inline-anonymous-functions` mode.
This also makes the generated code more aligned with modern JS.

This feature is enabled by default when targeting ES2015.
Alternatively, you can enable it by using the `-Xes-arrow-functions` command line argument.

> [Learn more about ES2015 (ECMAScript 2015, ES6) in the official documentation.](https://262.ecma-international.org/6.0/)

Reviewer: ```suggestion
[Learn more about ES2015 (ECMAScript 2015, ES6) in the official documentation](https://262.ecma-international.org/6.0/).
```

        with(compilerOptions) {
            if (this is KotlinJvmCompilerOptions) {
                jvmTarget.set(JvmTarget.JVM_17)
            }
        }
    }
}
```

This configures the JVM target to 17 for both JVM and Android projects.

To configure compiler options specifically for JVM projects, use `KotlinJvmExtension`:

```kotlin
configure<KotlinJvmExtension> {
    compilerOptions {
        jvmTarget.set(JvmTarget.JVM_17)
    }

    target.mavenPublication {
        groupId = "com.example"
        artifactId = "example-project"
        version = "1.0-SNAPSHOT"
    }
}
```

This example similarly configures the JVM target to 17 for JVM projects.
It also configures a Maven publication for the project so that its output is published to a Maven repository.

You can use the `KotlinAndroidExtension` in exactly the same way.

### Compiler symbols hidden from the Kotlin Gradle plugin API

Starting with Kotlin 2.1.0,
you will receive a warning if you access compiler module symbols bundled within the Kotlin Gradle plugin (KGP).
Previously, the KGP included `org.jetbrains.kotlin:kotlin-compiler-embeddable` in its runtime dependencies,

> making internal compiler symbols, such as `KotlinCompilerVersion`, available in build script classpath.

Reviewer: ```suggestion
making internal compiler symbols, such as `KotlinCompilerVersion`, available in the build script classpath.
```

        }
        workQueue.submit(ActionUsingKotlinCompiler::class.java) {}
    }
}
```

Finally, configure the Kotlin compiler classpath in your Gradle plugin:

```kotlin
import org.gradle.api.Plugin
import org.gradle.api.Project
abstract class MyPlugin: Plugin<Project> {
    override fun apply(target: Project) {
        val myDependencyScope = target.configurations.create("myDependencyScope")
        target.dependencies.add(myDependencyScope.name, "$KOTLIN_COMPILER_EMBEDDABLE:$KOTLIN_COMPILER_VERSION")
        val myResolvableConfiguration = target.configurations.create("myResolvable") {
            extendsFrom(myDependencyScope)
        }
        target.tasks.register("myTask", TaskUsingKotlinCompiler::class.java) {
            kotlinCompiler.from(myResolvableConfiguration)
        }
    }

    companion object {
        const val KOTLIN_COMPILER_EMBEDDABLE = "org.jetbrains.kotlin:kotlin-compiler-embeddable"
        const val KOTLIN_COMPILER_VERSION = "%kotlinVersion%"
    }
}
```

## Compose compiler updates

### Support for multiple stability configuration files

The Compose compiler can interpret multiple stability configuration files,

> but the `stabilityConfigurationFile` option of the Compose Compiler Gradle plugin previously allowed

Reviewer: ```suggestion
but the `stabilityConfigurationFile` option of the Compose Compiler Gradle plugin was previously allowed
```

Reviewer: We're missing a verb here. Does "was" work?

Author: Not sure, that it is necessary here.

  This change reflects the transition from the legacy memory manager in Kotlin/Native,
  which required freezing objects to share them between threads.
  To learn how to migrate from freezing-related APIs in the new memory model,
  see the [Kotlin/Native migration guide](native-migration-guide.md#update-your-code).
  For more information, see the [announcement about the deprecation of freezing](whatsnew1720.md#freezing).

* **`appendln()` is deprecated in favor of `appendLine()`:**
  The `StringBuilder.appendln()` and `Appendable.appendln()` functions are now deprecated, and using them results in an error.
  To replace them, use the `StringBuilder.appendLine()` or `Appendable.appendLine()` functions instead.
  The `appendln()` function is deprecated because, on Kotlin/JVM, it uses the `line.separator` system property,
  which has a different default value on each OS. On Kotlin/JVM, this property defaults to `\r\n` (CR LF) on Windows and `\n` (LF) on other systems.
  On the other hand, the `appendLine()` function consistently uses `\n` (LF) as the line separator, ensuring consistent behavior across platforms.

> For a complete list of affected APIs in this release, see [KT-71628](https://youtrack.jetbrains.com/issue/KT-71628).

Reviewer: ```suggestion
For a complete list of affected APIs in this release, see the [KT-71628](https://youtrack.jetbrains.com/issue/KT-71628) YouTrack issue.
```
